Add graph op replacement pass

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70396
* #68358
* __->__ #69915
* #69914

Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

tugsbayasgalan committed:
Add graph op replacement pass

[ghstack-poisoned]

tugsbayasgalan commented on test/cpp/jit/test_op_replacement.cpp:
```diff
@@ -0,0 +1,89 @@
+  EXPECT_EQ(g->get_op_version().value(), 4);
+}
+
+TEST(OpReplacementTest, ReplaceTestSubcmulInSimpleFunction) {
```
will add more tests once the adding testing utility methods PR is merged

tugsbayasgalan committed:
Update on "Add graph op replacement pass"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add graph op replacement pass"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add graph op replacement pass"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add graph op replacement pass"




[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198158).

tugsbayasgalan committed:
Update on "Add graph op replacement pass"


Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198158).

tugsbayasgalan committed:
Update on "Add graph op replacement pass"


Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198158).

tugsbayasgalan committed:
Update on "Add graph op replacement pass"


Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198158).

tugsbayasgalan committed:
Update on "Add graph op replacement pass"


Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198158).

gmagogsfm commented on torch/csrc/jit/passes/replacement_of_old_operators.cpp:
```diff
@@ -0,0 +1,89 @@
+              updated_version, entry[entry.size() - 1].bumped_at_version);
+          auto upgrader_entry =
+              findUpgrader(version_entry->second, current_version);
+          if (!upgrader_entry.has_value()) {
+            if (!isOpSymbolCurrent(schema_name, current_version)) {
+              TORCH_INTERNAL_ASSERT(
+                  false, "Upgrader must be present for ", schema_name);
```
When will this condition trigger? Like when we remove an old upgrader but then an even older op shows up in model?

tugsbayasgalan replied:
When we encounter a model where it has an operator with deprecated upgrader. We should probably want to error out in this case right?

gmagogsfm replied:
Yes, we should. The error message needs to be more verbose though.

tugsbayasgalan replied:
Done

gmagogsfm commented on torch/csrc/jit/passes/replacement_of_old_operators.cpp:
```diff
@@ -0,0 +1,89 @@
+              TORCH_INTERNAL_ASSERT(
+                  false, "Upgrader must be present for ", schema_name);
+            }
+            return;
```
Should this be `continue` after moving node to next?

tugsbayasgalan replied:
Yeah that is true. I was thinking if we see an up-to-date version of the operator, we shouldn't bother applying upgraders anymore. But yeah, it might not be the case for other operators. It should indeed be `continue`

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/passes/replacement_of_old_operators.cpp:
```diff
@@ -0,0 +1,89 @@
+              upgrader_name,
+              " must exist");
+          Graph upgrader_graph;
+          parseIR(upgrader_graph_entry->second, &upgrader_graph);
```
Could we cache parsed IR? It seems a waste of cycles to parse for every instance of old op invocation.

tugsbayasgalan replied:
Sounds good.

tugsbayasgalan replied:
done

tugsbayasgalan commented on torch/csrc/jit/passes/replacement_of_old_operators.cpp:
```diff
@@ -0,0 +1,89 @@
+
+    // now that we updated the graph, we want to bump the
+    // graph version too.
+    graph_->set_op_version(updated_version);
```
I thought setting to kProduce would be too conservative. If the graph has operators that were updated a while ago (say 4) , it is probably better to have the model version around that instead of the current server version (say 100) because the older runtimes that are less 100 can still run the model.

tugsbayasgalan replied:
resolved

tugsbayasgalan commented on torch/csrc/jit/passes/replacement_of_old_operators.h:
```diff
@@ -0,0 +1,11 @@
+#pragma once
```
The reason for this was it seemed like all the pass file names are nouns and the pass itself has verb name.

tugsbayasgalan replied:
Renamed the pass

tugsbayasgalan committed:
Update on "Add graph op replacement pass"


Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add graph op replacement pass"


Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198158).

tugsbayasgalan commented on torch/csrc/jit/passes/replacement_of_old_operators.h:
```diff
@@ -0,0 +1,11 @@
+namespace torch {
+namespace jit {
+
+TORCH_API void ApplyOldOpsUpgraders(std::shared_ptr<Graph> graph);
```
done

gmagogsfm commented on torch/csrc/jit/passes/replacement_of_old_operators.cpp:
```diff
@@ -0,0 +1,96 @@
+            upgrader_graph = upgrader_graph_it->second;
+          } else {
+            parseIR(upgrader_graph_entry->second, upgrader_graph.get());
+            upgraderCache[upgrader_name] = upgrader_graph;
```
It would be cleaner if we devise a helper function to read/update upgraderCache in a single call instead of complicating core logic of this pass with cache manipulations. Additionally, we can even compile all upgraders as first step of this pass for future uses.

tugsbayasgalan replied:
Moved it into an utility function.

tugsbayasgalan committed:
Update on "Add graph op replacement pass"


Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198158).

tugsbayasgalan committed:
Update on "Add graph op replacement pass"


Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198158).

gmagogsfm commented on test/cpp/jit/test_op_replacement.cpp:
```diff
@@ -0,0 +1,145 @@
+  ReplaceOldOperatorsWithUpgraders(g);
+  testing::FileCheck().check_count("aten::mul", 1, false)->run(*g);
+
+  testing::FileCheck().check_count("aten::sub", 1, false)->run(*g);
```
Consider checking `_test_serialization_subcmul` is not present in the graph any more

gmagogsfm commented on torch/csrc/jit/passes/replacement_of_old_operators.cpp:
```diff
@@ -0,0 +1,110 @@
+  auto upgrader_graph_entry = dump_upgraders_map().find(upgrader_name);
+  TORCH_INTERNAL_ASSERT(
+      upgrader_graph_entry != dump_upgraders_map().end(),
+      "Corresponding upgrader graph for ",
```
Users might not be aware of what an upgrader is. Can we frame it as "Op X at older version Y is out of support window, please regenerate your model with newer version of PyTorch"

gmagogsfm replied:
Same for the warning on line 66

gmagogsfm reviewed (approved):
Looks good, just some minor comments

tugsbayasgalan committed:
Update on "Add graph op replacement pass"


Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198158).

tugsbayasgalan committed:
Update on "Add graph op replacement pass"


Differential Revision: [D33198158](https://our.internmc.facebook.com/intern/diff/D33198158)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198158).
